src/main/scala/mesosphere/marathon/state/Group.scala

```diff
@@ -152,32 +153,79 @@ object Group extends StrictLogging {
   def validGroup(base: PathId, enabledFeatures: Set[String]): Validator[Group] =
     validator[Group] { group =>
       group.id is validPathWithBase(base)
-      group.apps.values as "apps" is every(
-        AppDefinition.validNestedAppDefinition(group.id.canonicalPath(base), enabledFeatures))
+
+      group.transitiveApps as "apps" is everyApp(
+        AppDefinition.validBasicAppDefinition(enabledFeatures) and isChildOfParentId(group)
+      )
+
       group is noAppsAndPodsWithSameId
       group is noAppsAndGroupsWithSameName
       group is noPodsAndGroupsWithSameName
-      group.groupsById.values as "groups" is every(validGroup(group.id.canonicalPath(base), enabledFeatures))
+
+      group.transitiveGroupValues as "groups" is every(
+        noAppsAndPodsWithSameId and noAppsAndGroupsWithSameName and noPodsAndGroupsWithSameName
+          and isTrue("Group has to be child of groups with parent id") { childGroup =>
+            if (childGroup.id.parent.isRoot) group.groupsById.contains(childGroup.id)
+            else {
+              group.group(childGroup.id.parent) match {
+                case None => false
+                case Some(parentGroup) => parentGroup.groupsById.contains(childGroup.id)
+              }
+            }
+          }
+      )
     }
 
+  implicit def everyApp(validator: Validator[AppDefinition]): Validator[Iterable[AppDefinition]] = {
+    new Validator[Iterable[AppDefinition]] {
+      override def apply(seq: Iterable[AppDefinition]): Result = {
+        seq.foldLeft[Result](Success) {
+          case (accum, (app)) =>
+            validator(app) match {
+              case Success => accum
+              case Failure(violations) =>
+                val scopedViolations = violations.map { violation =>
+                  violation.withPath(Descriptions.Explicit(app.id.toString))
+                }
+                accum.and(Failure(scopedViolations))
+            }
+        }
+      }
+    }
+  }
+
+  private def isChildOfParentId(group: Group): Validator[AppDefinition] = {
+    isTrue("App has to be child of group with parent id") { app =>
+      if (app.id.parent.isRoot) group.apps.contains(app.id)
+      else {
+        group.group(app.id.parent) match {
```

Comment:
Shouldn't it be smth. like `group.group(app.id.parent).id == group.id`? (just checking the ids)

Reply:
I've changed it to `(app.id.parent == group.id) group.apps.contains(app.id)`

Reply:
On the second thought shouldn't we check the equality of the whole group e.g. `app.id.parent == group`?